src/server/database/Database/PreparedStatement.cpp

```diff
@@ -36,51 +36,39 @@ void PreparedStatement::BindParameters(MySQLPreparedStatement* stmt)
-                break;
-            case TYPE_UI8:
-                stmt->setUInt8(pos, std::get<uint8>(data.data));
-                break;
-            case TYPE_UI16:
-                stmt->setUInt16(pos, std::get<uint16>(data.data));
-                break;
-            case TYPE_UI32:
-                stmt->setUInt32(pos, std::get<uint32>(data.data));
-                break;
-            case TYPE_I8:
-                stmt->setInt8(pos, std::get<int8>(data.data));
-                break;
-            case TYPE_I16:
-                stmt->setInt16(pos, std::get<int16>(data.data));
-                break;
-            case TYPE_I32:
-                stmt->setInt32(pos, std::get<int32>(data.data));
-                break;
-            case TYPE_UI64:
-                stmt->setUInt64(pos, std::get<uint64>(data.data));
-                break;
-            case TYPE_I64:
-                stmt->setInt64(pos, std::get<int64>(data.data));
-                break;
-            case TYPE_FLOAT:
-                stmt->setFloat(pos, std::get<float>(data.data));
-                break;
-            case TYPE_DOUBLE:
-                stmt->setDouble(pos, std::get<double>(data.data));
-                break;
-            case TYPE_STRING:
-                stmt->setString(pos, std::get<std::string>(data.data));
-                break;
-            case TYPE_BINARY:
-                stmt->setBinary(pos, std::get<std::vector<uint8>>(data.data));
-                break;
-            case TYPE_NULL:
+        std::visit([&](auto&& data) {
+            using T = std::decay_t<decltype(data)>;
+            if constexpr (std::is_same_v<T, bool>)
+                stmt->setBool(pos, data);
+            else if constexpr (std::is_same_v<T, uint8>)
+                stmt->setUInt8(pos, data);
+            else if constexpr (std::is_same_v<T, uint16>)
+                stmt->setUInt16(pos, data);
+            else if constexpr (std::is_same_v<T, uint32>)
+                stmt->setUInt32(pos, data);
+            else if constexpr (std::is_same_v<T, uint64>)
+                stmt->setUInt64(pos, data);
+            else if constexpr (std::is_same_v<T, int8>)
+                stmt->setInt8(pos, data);
+            else if constexpr (std::is_same_v<T, int16>)
+                stmt->setInt16(pos, data);
+            else if constexpr (std::is_same_v<T, int32>)
+                stmt->setInt32(pos, data);
+            else if constexpr (std::is_same_v<T, int64>)
+                stmt->setInt64(pos, data);
+            else if constexpr (std::is_same_v<T, float>)
+                stmt->setFloat(pos, data);
+            else if constexpr (std::is_same_v<T, double>)
+                stmt->setDouble(pos, data);
+            else if constexpr (std::is_same_v<T, std::string>)
+                stmt->setString(pos, data);
+            else if constexpr (std::is_same_v<T, std::vector<uint8>>)
+                stmt->setBinary(pos, data);
```

Comment:
`stmt->setBinary(pos, data);` -> `stmt->setBinary(pos, std::forward<decltype(data)>(data));`

Reply:
May i ask why this and the string method should have this whereas you don't want it for the primitive types?

Reply:
You can do it for the other types as well but it won't bring any benefit since you would prefer copying the smaller types sizeof(type)<= sizeof(void*) anyway. If you miss the perfect forwarding you will create a copy of the string or vector.

Also don't forget to do

```cpp
std::visit(..., std::move(data));
//                     ^^^^^
```

if you intend to fully consume the variant through the visit.

Reply:
@Naios Are you on IRC or Discord? I still don't really see the thing considering we're using auto&& in the visitor, discussion is a bit hard here and I'd like to learn.

Using I'm unsure about when we're finished with the variant or not, I don't know how we should proper safeguard never fetching the querystring twice. Or am i missunderstanding something here?

Thanks :smile: